# Car Price Prediction

## Importing standard libraries

In [631]:
import numpy as np
import pandas as pd

## Data exploration

### Loading the dataset
as a pandas DataFrame

In [632]:
dataset = pd.read_csv('./assets/car-details.csv')

Retrieving some information about the dataset

In [633]:
dataset.head(8)

name  year  selling_price  km_driven    fuel  \
0         Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1   Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2       Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3      Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4         Maruti Swift VXI BSIII  2007         130000     120000  Petrol   
5  Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000  Petrol   
6   Maruti Wagon R LXI DUO BSIII  2007          96000     175000     LPG   
7             Maruti 800 DX BSII  2001          45000       5000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   
5  Individual       Manual   First Owner  20.14 kmpl  1197 CC   81.86 bhp   
6  Individual       Manual   First Owner  17.3 km/kg  1061 CC    57.5 bhp   
7  Individual       Manual  Second Owner   16.1 kmpl   796 CC      37 bhp   

                     torque  seats  
0            190Nm@ 2000rpm    5.0  
1       250Nm@ 1500-2500rpm    5.0  
2     12.7@ 2,700(kgm@ rpm)    5.0  
3  22.4 kgm at 1750-2750rpm    5.0  
4     11.5@ 4,500(kgm@ rpm)    5.0  
5         113.75nm@ 4000rpm    5.0  
6      7.8@ 4,500(kgm@ rpm)    5.0  
7             59Nm@ 2500rpm    4.0

Having seen the shape of the dataset, it is time to see if some of the values are missing.

In [634]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB


In [635]:
dataset.isna().sum()

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        215
torque           222
seats            221
dtype: int64

In the output we can see that the dataset has a total of 8128 rows and 13 columns. 

Also we can see that the dataset has some missing values in the columns mileage, engine, max_power, torque and seats.

**Now I will go through the columns we believe to be categorical**

In [636]:
print("Column: Fuel types")
print(dataset.fuel.value_counts())

print("\nColumn: Seller types")
print(dataset.seller_type.value_counts())

print("\nColumn: Transmission types")
print(dataset.transmission.value_counts())

print("\nColumn: Owner count")
print(dataset.owner.value_counts())

Column: Fuel types
Diesel    4402
Petrol    3631
CNG         57
LPG         38
Name: fuel, dtype: int64

Column: Seller types
Individual          6766
Dealer              1126
Trustmark Dealer     236
Name: seller_type, dtype: int64

Column: Transmission types
Manual       7078
Automatic    1050
Name: transmission, dtype: int64

Column: Owner count
First Owner             5289
Second Owner            2105
Third Owner              555
Fourth & Above Owner     174
Test Drive Car             5
Name: owner, dtype: int64


## Cleaning up the data

Okay so having seen the and understood the plan. Here is the TODO to clean up the data.

* Take care of the missing values
* Removing the car model from the brand name
* Convert max_power, milage and engine to numeric values
* Extract torque to a usable value
* Deal with outliers
* Exporting the dataset for visualization
* Encode the categorical values
* Exporting the dataset

### Taking care of missing values

Going back to see the isna() function. It seems a lot of the missing values are on the same row of an entry

Therefore we can use the dropna() function to drop the rows with missing values. As it doesn't seem to make sense to try to predict

In [637]:
dataset = dataset.dropna()

In [638]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7906 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           7906 non-null   object 
 1   year           7906 non-null   int64  
 2   selling_price  7906 non-null   int64  
 3   km_driven      7906 non-null   int64  
 4   fuel           7906 non-null   object 
 5   seller_type    7906 non-null   object 
 6   transmission   7906 non-null   object 
 7   owner          7906 non-null   object 
 8   mileage        7906 non-null   object 
 9   engine         7906 non-null   object 
 10  max_power      7906 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7906 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 864.7+ KB


In [639]:
dataset.isna().sum()

name             0
year             0
selling_price    0
km_driven        0
fuel             0
seller_type      0
transmission     0
owner            0
mileage          0
engine           0
max_power        0
torque           0
seats            0
dtype: int64

In [640]:
dataset.reset_index(drop=True, inplace=True)

Now the missing data has been taken care of. If the missing data was more than a few rows, it would be a good idea to try to impute the missing values.

However in this case, it simply didn't make any sense to do so.

### Removing the car model from the brand name

I'm afraid keeping the exact car model, will make it difficult to predict the price of not recorded vehicles.

However keeping the brand name can still be useful, as it can indicate luxury of a car.

In [641]:
brands = dataset.name.copy()
dataset = dataset.drop(['name'], axis=1)

In [642]:
for i in range(len(brands)):
    brands[i] = brands[i].split(' ')[0]

Get the amount of unique brand names

In [643]:
len(brands.unique())

31

In [644]:
brands.value_counts()

Maruti           2367
Hyundai          1360
Mahindra          758
Tata              719
Honda             466
Toyota            452
Ford              388
Chevrolet         230
Renault           228
Volkswagen        185
BMW               118
Skoda             104
Nissan             81
Jaguar             71
Volvo              67
Datsun             65
Mercedes-Benz      54
Fiat               41
Audi               40
Lexus              34
Jeep               31
Mitsubishi         14
Force               6
Land                6
Isuzu               5
Kia                 4
Ambassador          4
Daewoo              3
MG                  3
Ashok               1
Opel                1
Name: name, dtype: int64

In [645]:
dataset['brand'] = brands

Rearranging the columns for clarity

In [646]:
dataset_cols = ['brand', 'year', 'km_driven', 'fuel', 'transmission', 'seller_type', 'owner', 'seats', 'mileage', 'engine', 'max_power', 'torque', 'selling_price']

dataset = dataset[dataset_cols]

In [647]:
dataset.head()

brand  year  km_driven    fuel transmission seller_type         owner  \
0   Maruti  2014     145500  Diesel       Manual  Individual   First Owner   
1    Skoda  2014     120000  Diesel       Manual  Individual  Second Owner   
2    Honda  2006     140000  Petrol       Manual  Individual   Third Owner   
3  Hyundai  2010     127000  Diesel       Manual  Individual   First Owner   
4   Maruti  2007     120000  Petrol       Manual  Individual   First Owner   

   seats     mileage   engine   max_power                    torque  \
0    5.0   23.4 kmpl  1248 CC      74 bhp            190Nm@ 2000rpm   
1    5.0  21.14 kmpl  1498 CC  103.52 bhp       250Nm@ 1500-2500rpm   
2    5.0   17.7 kmpl  1497 CC      78 bhp     12.7@ 2,700(kgm@ rpm)   
3    5.0   23.0 kmpl  1396 CC      90 bhp  22.4 kgm at 1750-2750rpm   
4    5.0   16.1 kmpl  1298 CC    88.2 bhp     11.5@ 4,500(kgm@ rpm)   

   selling_price  
0         450000  
1         370000  
2         158000  
3         225000  
4         130000

**Succes!** The car brand names are now added to the dataset, and should be used as a categorical feature.

### Converting max_power, milage and engine to numeric values

Transforming the columns into numeric values, so they can be used by the machine learning models

In [648]:
columns = ["mileage", "engine", "max_power"]

for column in columns:
    dataset[column] = dataset[column].str.replace(',', '').str.replace('kmpl', '').str.replace('km/kg', '').str.replace('CC', '').str.replace('bhp', '')
    dataset[column] = pd.to_numeric(dataset[column])

print(dataset[columns].isna().sum())
print(dataset[columns].info())

mileage      0
engine       0
max_power    0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7906 entries, 0 to 7905
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   mileage    7906 non-null   float64
 1   engine     7906 non-null   int64  
 2   max_power  7906 non-null   float64
dtypes: float64(2), int64(1)
memory usage: 185.4 KB
None


Perfect! Now we are getting somewhere. The dataset only needs a few more columns altered to be useable.

### Extracting torque to a usable value

The plan to handle this is to extract the nm from the torque column and then saving it as a different column

In [649]:
import re

nm = dataset.torque.copy()

def convert_kgm_to_nm(x):
    return np.round(x * 9.80665, decimals=2)

for i in range(len(nm)):
    nm[i] = nm[i].replace(",", ".")
    torque = float(re.findall(r'\d+\.?\d*', nm[i])[0])

    if nm[i].lower().find('kgm') != -1 and torque < 90:
        nm[i] = float(convert_kgm_to_nm(torque))
    elif torque > 780: # Maruti Zen D has a massive outlier in the dataset
        nm[i] = float(78.9)
    else:
        nm[i] = torque

print('Min: ', nm.min()) # Tata Nano 
print('Max: ', nm.max()) # Volvo XC90
print('Median: ', nm.median())

Min:  47.07
Max:  640.0
Median:  170.0


Removing the torque column and adding the nm column (not really necessary for predicting the price, but will be for fuel type classification)

In [650]:
nm = pd.to_numeric(nm, errors='coerce')

dataset['nm'] = nm
dataset.drop(['torque'], axis=1)

dataset_cols.insert(dataset_cols.index('torque'), 'nm')
dataset_cols.remove('torque')

dataset = dataset[dataset_cols]
dataset.head()

brand  year  km_driven    fuel transmission seller_type         owner  \
0   Maruti  2014     145500  Diesel       Manual  Individual   First Owner   
1    Skoda  2014     120000  Diesel       Manual  Individual  Second Owner   
2    Honda  2006     140000  Petrol       Manual  Individual   Third Owner   
3  Hyundai  2010     127000  Diesel       Manual  Individual   First Owner   
4   Maruti  2007     120000  Petrol       Manual  Individual   First Owner   

   seats  mileage  engine  max_power      nm  selling_price  
0    5.0    23.40    1248      74.00  190.00         450000  
1    5.0    21.14    1498     103.52  250.00         370000  
2    5.0    17.70    1497      78.00  124.54         158000  
3    5.0    23.00    1396      90.00  219.67         225000  
4    5.0    16.10    1298      88.20  112.78         130000

### Taking care of massive outliers in the 95th percentile

This removes the outliers from 0-5% and 95-100% of the dataset. Majority of the time, these are outliers skewed by wrong data.

I don't understand the inner workings of this, but it works. Thanks Stack Overflow!

In [651]:
cols = ['km_driven', 'seats', 'mileage', 'engine', 'max_power', 'nm', 'selling_price'] # one or more

lower = dataset[cols].quantile(0.05)
upper = dataset[cols].quantile(0.95)
IQR = upper - lower

dataset = dataset[~((dataset[cols] < (lower - 1.5 * IQR)) |(dataset[cols] > (upper + 1.5 * IQR))).any(axis=1)]
dataset.reset_index(drop=True, inplace=True)

dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7776 entries, 0 to 7775
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   brand          7776 non-null   object 
 1   year           7776 non-null   int64  
 2   km_driven      7776 non-null   int64  
 3   fuel           7776 non-null   object 
 4   transmission   7776 non-null   object 
 5   seller_type    7776 non-null   object 
 6   owner          7776 non-null   object 
 7   seats          7776 non-null   float64
 8   mileage        7776 non-null   float64
 9   engine         7776 non-null   int64  
 10  max_power      7776 non-null   float64
 11  nm             7776 non-null   float64
 12  selling_price  7776 non-null   int64  
dtypes: float64(4), int64(4), object(5)
memory usage: 789.9+ KB


### Exporting the dataset at this point for data visualization

We want to export the dataset at this point before encoding the categorical values to make it easier to visualize the data.

In [652]:
dataset.to_csv('./assets/car-details-for-visualization.csv', index=False)

### Encode the categorical values

Now it's time to encode the categorical values. The values that will be converted are:
* Fuel type
* Seller type
* Transmission
* Owner
* Brand

**Transmission types**

Since the transmission types only alter between manual and automatic, they can be binarized.

In [653]:
print("\nColumn: Transmission types")
print(dataset.transmission.value_counts())


Column: Transmission types
Manual       6852
Automatic     924
Name: transmission, dtype: int64


In [654]:
# Used label_binarize() instead of LabelBinarizer() since it would allow me to have automatic transmission as 1 instead of 0

from sklearn.preprocessing import label_binarize

dataset[['transmission']] = label_binarize(dataset[['transmission']], classes=['Manual', 'Automatic'])

dataset.transmission.value_counts()

0    6852
1     924
Name: transmission, dtype: int64

**Seller types, Owner & Fuel types**

I think there is a strong relation between the seller types, Fuel types and amount of owners in relation to the selling price, so I will label/ordinal encode them

In [655]:
print("\nColumn: Seller types")
print(dataset.seller_type.value_counts())

print("\nColumn: Owner count")
print(dataset.owner.value_counts())

print("Column: Fuel types")
print(dataset.fuel.value_counts())


Column: Seller types
Individual          6543
Dealer               997
Trustmark Dealer     236
Name: seller_type, dtype: int64

Column: Owner count
First Owner             5096
Second Owner            2009
Third Owner              509
Fourth & Above Owner     160
Test Drive Car             2
Name: owner, dtype: int64
Column: Fuel types
Diesel    4212
Petrol    3477
CNG         52
LPG         35
Name: fuel, dtype: int64


In [656]:
from sklearn.preprocessing import LabelEncoder

label_enc = LabelEncoder()
dataset['seller_type'] = label_enc.fit_transform(dataset['seller_type'])
dataset['owner'] = label_enc.fit_transform(dataset['owner'])
dataset['fuel'] = label_enc.fit_transform(dataset['fuel'])

dataset.head(8)

brand  year  km_driven  fuel  transmission  seller_type  owner  seats  \
0   Maruti  2014     145500     1             0            1      0    5.0   
1    Skoda  2014     120000     1             0            1      2    5.0   
2    Honda  2006     140000     3             0            1      4    5.0   
3  Hyundai  2010     127000     1             0            1      0    5.0   
4   Maruti  2007     120000     3             0            1      0    5.0   
5  Hyundai  2017      45000     3             0            1      0    5.0   
6   Maruti  2007     175000     2             0            1      0    5.0   
7   Maruti  2001       5000     3             0            1      2    4.0   

   mileage  engine  max_power      nm  selling_price  
0    23.40    1248      74.00  190.00         450000  
1    21.14    1498     103.52  250.00         370000  
2    17.70    1497      78.00  124.54         158000  
3    23.00    1396      90.00  219.67         225000  
4    16.10    1298      88.20  112.78         130000  
5    20.14    1197      81.86  113.75         440000  
6    17.30    1061      57.50   76.49          96000  
7    16.10     796      37.00   59.00          45000

Before one hot encoding the brands, I would like to export the dataset as it is for data visualization.

In [657]:
dataset.to_csv('./assets/car-details-for-visualization.csv', index=False)

**Brands**

There might be a small ordinal relation between the brands, but I don't think it's enough. Therefore I will One Hot encode them

In [658]:
one_hot_brand = pd.get_dummies(dataset.brand)

dataset = dataset.drop(['brand'], axis=1)
dataset = dataset.join(one_hot_brand)

dataset.head()

year  km_driven  fuel  transmission  seller_type  owner  seats  mileage  \
0  2014     145500     1             0            1      0    5.0    23.40   
1  2014     120000     1             0            1      2    5.0    21.14   
2  2006     140000     3             0            1      4    5.0    17.70   
3  2010     127000     1             0            1      0    5.0    23.00   
4  2007     120000     3             0            1      0    5.0    16.10   

   engine  max_power  ...  Mercedes-Benz  Mitsubishi  Nissan  Opel  Renault  \
0    1248      74.00  ...              0           0       0     0        0   
1    1498     103.52  ...              0           0       0     0        0   
2    1497      78.00  ...              0           0       0     0        0   
3    1396      90.00  ...              0           0       0     0        0   
4    1298      88.20  ...              0           0       0     0        0   

   Skoda  Tata  Toyota  Volkswagen  Volvo  
0      0     0       0           0      0  
1      1     0       0           0      0  
2      0     0       0           0      0  
3      0     0       0           0      0  
4      0     0       0           0      0  

[5 rows x 42 columns]

In [659]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7776 entries, 0 to 7775
Data columns (total 42 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           7776 non-null   int64  
 1   km_driven      7776 non-null   int64  
 2   fuel           7776 non-null   int32  
 3   transmission   7776 non-null   int32  
 4   seller_type    7776 non-null   int32  
 5   owner          7776 non-null   int32  
 6   seats          7776 non-null   float64
 7   mileage        7776 non-null   float64
 8   engine         7776 non-null   int64  
 9   max_power      7776 non-null   float64
 10  nm             7776 non-null   float64
 11  selling_price  7776 non-null   int64  
 12  Ambassador     7776 non-null   uint8  
 13  Ashok          7776 non-null   uint8  
 14  Audi           7776 non-null   uint8  
 15  BMW            7776 non-null   uint8  
 16  Chevrolet      7776 non-null   uint8  
 17  Daewoo         7776 non-null   uint8  
 18  Datsun  

Now all the categorical values are encoded and we are left with a 47 columns dataset.

All there is left to do is moving the dependent variable to the end of the dataset.

In [660]:
dep_variable = dataset.pop('selling_price')

dataset.insert(len(dataset.columns), 'selling_price', dep_variable)

dataset.head(8)

year  km_driven  fuel  transmission  seller_type  owner  seats  mileage  \
0  2014     145500     1             0            1      0    5.0    23.40   
1  2014     120000     1             0            1      2    5.0    21.14   
2  2006     140000     3             0            1      4    5.0    17.70   
3  2010     127000     1             0            1      0    5.0    23.00   
4  2007     120000     3             0            1      0    5.0    16.10   
5  2017      45000     3             0            1      0    5.0    20.14   
6  2007     175000     2             0            1      0    5.0    17.30   
7  2001       5000     3             0            1      2    4.0    16.10   

   engine  max_power  ...  Mitsubishi  Nissan  Opel  Renault  Skoda  Tata  \
0    1248      74.00  ...           0       0     0        0      0     0   
1    1498     103.52  ...           0       0     0        0      1     0   
2    1497      78.00  ...           0       0     0        0      0     0   
3    1396      90.00  ...           0       0     0        0      0     0   
4    1298      88.20  ...           0       0     0        0      0     0   
5    1197      81.86  ...           0       0     0        0      0     0   
6    1061      57.50  ...           0       0     0        0      0     0   
7     796      37.00  ...           0       0     0        0      0     0   

   Toyota  Volkswagen  Volvo  selling_price  
0       0           0      0         450000  
1       0           0      0         370000  
2       0           0      0         158000  
3       0           0      0         225000  
4       0           0      0         130000  
5       0           0      0         440000  
6       0           0      0          96000  
7       0           0      0          45000  

[8 rows x 42 columns]

Great! The dataset is ready to be used by the machine learning models. All there is left to do is exporting the dataset.

In [661]:
dataset.to_csv('./assets/car-details-for-ml.csv', index=False)